### DISCUSSION TITLE
While using Oculus Quest 2 controllers, ControllerModelLoader says "Unable to load model animation."

### DISCUSSION THREAD

**Discussion by s871922003** [microsoft/MixedRealityToolkit-Unity#11478] [Q&A]
Hello, I was trying to run the HandInteractionExample scene with Oculus Quest 2, but something went wrong and ControllerModelLoader.cs throw out an error log that says "Unable to load model animation".

This results in a visual mode that despite having the correct shape of the Quest 2 controllers but the whole thing only has bright pink color on it.

The logs I get from running **HandInteractionExample** scene are as followed:
2023/04/21 16:27:36.663 7139 7139 Warn UnityMain type=1400 audit(0.0:118): avc: denied { read } for name="u:object_r:vendor_board_init_prop:s0" dev="tmpfs" ino=13457 scontext=u:r:untrusted_app_29:s0:c92,c256,c512,c768 tcontext=u:object_r:vendor_board_init_prop:s0 tclass=file permissive=0 app=com.Microsoft.MRTK3BluetoothDemo
2023/04/21 16:27:36.683 7139 7139 Warn UnityMain type=1400 audit(0.0:119): avc: denied { read } for name="u:object_r:vendor_board_init_prop:s0" dev="tmpfs" ino=13457 scontext=u:r:untrusted_app_29:s0:c92,c256,c512,c768 tcontext=u:object_r:vendor_board_init_prop:s0 tclass=file permissive=0 app=com.Microsoft.MRTK3BluetoothDemo
2023/04/21 16:27:36.713 7139 7139 Warn UnityMain type=1400 audit(0.0:120): avc: denied { read } for name="u:object_r:vendor_board_init_prop:s0" dev="tmpfs" ino=13457 scontext=u:r:untrusted_app_29:s0:c92,c256,c512,c768 tcontext=u:object_r:vendor_board_init_prop:s0 tclass=file permissive=0 app=com.Microsoft.MRTK3BluetoothDemo
2023/04/21 16:27:36.746 7139 7156 Warn Unity Orientation Listener cannot detect orientation.
2023/04/21 16:27:41.353 7139 7156 Error Unity DllNotFoundException: Unable to load DLL 'AudioPluginOculusSpatializer': The specified module could not be found.
2023/04/21 16:27:41.353 7139 7156 Error Unity   at ONSPAudioSource.OSP_SetGlobalVoiceLimit (System.Int32 VoiceLimit) [0x00000] in <00000000000000000000000000000000>:0 
2023/04/21 16:27:41.353 7139 7156 Error Unity 
2023/04/21 16:27:41.697 7139 7156 Warn Unity Camera "Main Camera" does not use a AR Pose Driver or a Tracked Pose Driver, so its transform will not be updated by an XR device.  In order for this to be updated, please add either an AR Pose Driver or a Tracked Pose Driver.
2023/04/21 16:27:42.081 7139 7156 Error Unity Unable to load model animation.
2023/04/21 16:27:42.081 7139 7156 Error Unity Microsoft.MixedReality.Toolkit.Input.<TryGenerateControllerModelFromPlatformSDK>d__3:MoveNext()
2023/04/21 16:27:42.081 7139 7156 Error Unity System.Threading.ContextCallback:Invoke(Object)
2023/04/21 16:27:42.081 7139 7156 Error Unity System.Threading.ExecutionContext:RunInternal(ExecutionContext, ContextCallback, Object, Boolean)
2023/04/21 16:27:42.081 7139 7156 Error Unity System.Runtime.CompilerServices.MoveNextRunner:Run()
2023/04/21 16:27:42.081 7139 7156 Error Unity System.Action:Invoke()
2023/04/21 16:27:42.081 7139 7156 Error Unity System.Threading.ContextCallback:Invoke(Object)
2023/04/21 16:27:42.081 7139 7156 Error Unity System.Threading.Tasks.AwaitTaskContinuation:RunCallback(ContextCallback, Object, Task&)
2023/04/21 16:27:42.081 7139 7156 Error Unity System.Threading.Tasks.Task:FinishContinuations()
2023/04/21 16:27:42.081 7139 7156 Error Unity System.Threading.Tasks.Task`1:TrySetResult(TResult)
2023/04/21 16:27:42.081 7139 7156 Error Unity System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1:SetResult(TResult)
2023/04/21 16:27:42.081 7139 7156 Error Unity GLTFast.<LoadGltfBinary>d__53:MoveNext()
2023/04/21 16:27:42.081 7139 7156 Error Unity System.Threading.ContextCallback:Invoke(Object)
2023/04/21 16:27:42.081 7139 7156 Error Unity System.Threading.ExecutionContext:RunInternal(ExecutionContext, ContextCallback, Object, Boolean)
2023/04/21 16:27:42.081 7139 7156 Error Unity System.Runtime.CompilerServices.MoveNextRunner:Run()
2023/04/21 16:27:42.081 7139 7156 Error Unity System.Action:Invoke()
2023/04/21 16:27:42.081 7139 7156 Error Unity System.Threading.ContextCallback:Invoke(Object)
2023/04/21 16:27:42.081 7139 7156 Error Unity System.Threading.Tasks.AwaitTaskContinuation:Ru

Following up on this one, I found that the model can be correctly displayed in the HandInteractionExample scene (with no animation still, but the visual is correct), but will have that pink appearance I mentioned in the SampleEmptyMRTKScene

The logs I get from running **SampleEmptyMRTKScene** are as followed:
2023/04/21 16:17:24.623 6307 6307 Warn UnityMain type=1400 audit(0.0:98): avc: denied { read } for name="u:object_r:vendor_board_init_prop:s0" dev="tmpfs" ino=13457 scontext=u:r:untrusted_app_29:s0:c92,c256,c512,c768 tcontext=u:object_r:vendor_board_init_prop:s0 tclass=file permissive=0 app=com.Microsoft.MRTK3BluetoothDemo
2023/04/21 16:17:24.643 6307 6307 Warn UnityMain type=1400 audit(0.0:99): avc: denied { read } for name="u:object_r:vendor_board_init_prop:s0" dev="tmpfs" ino=13457 scontext=u:r:untrusted_app_29:s0:c92,c256,c512,c768 tcontext=u:object_r:vendor_board_init_prop:s0 tclass=file permissive=0 app=com.Microsoft.MRTK3BluetoothDemo
2023/04/21 16:17:24.683 6307 6307 Warn UnityMain type=1400 audit(0.0:100): avc: denied { read } for name="u:object_r:vendor_board_init_prop:s0" dev="tmpfs" ino=13457 scontext=u:r:untrusted_app_29:s0:c92,c256,c512,c768 tcontext=u:object_r:vendor_board_init_prop:s0 tclass=file permissive=0 app=com.Microsoft.MRTK3BluetoothDemo
2023/04/21 16:17:24.731 6307 6324 Warn Unity Orientation Listener cannot detect orientation.
2023/04/21 16:17:24.912 6307 6611 Warn Unity The referenced script on this Behaviour (Game Object 'Input Tracker') is missing!
2023/04/21 16:17:29.346 6307 6324 Error Unity DllNotFoundException: Unable to load DLL 'AudioPluginOculusSpatializer': The specified module could not be found.
2023/04/21 16:17:29.346 6307 6324 Error Unity   at ONSPAudioSource.OSP_SetGlobalVoiceLimit (System.Int32 VoiceLimit) [0x00000] in <00000000000000000000000000000000>:0 
2023/04/21 16:17:29.346 6307 6324 Error Unity 
2023/04/21 16:17:29.362 6307 6324 Warn Unity Camera "Main Camera" does not use a AR Pose Driver or a Tracked Pose Driver, so its transform will not be updated by an XR device.  In order for this to be updated, please add either an AR Pose Driver or a Tracked Pose Driver.
2023/04/21 16:17:29.567 6307 6324 Error Unity Unable to load model animation.
2023/04/21 16:17:29.567 6307 6324 Error Unity Microsoft.MixedReality.Toolkit.Input.<TryGenerateControllerModelFromPlatformSDK>d__3:MoveNext()
2023/04/21 16:17:29.567 6307 6324 Error Unity System.Threading.ContextCallback:Invoke(Object)
2023/04/21 16:17:29.567 6307 6324 Error Unity System.Threading.ExecutionContext:RunInternal(ExecutionContext, ContextCallback, Object, Boolean)
2023/04/21 16:17:29.567 6307 6324 Error Unity System.Runtime.CompilerServices.MoveNextRunner:Run()
2023/04/21 16:17:29.567 6307 6324 Error Unity System.Action:Invoke()
2023/04/21 16:17:29.567 6307 6324 Error Unity System.Threading.ContextCallback:Invoke(Object)
2023/04/21 16:17:29.567 6307 6324 Error Unity System.Threading.Tasks.AwaitTaskContinuation:RunCallback(ContextCallback, Object, Task&)
2023/04/21 16:17:29.567 6307 6324 Error Unity System.Threading.Tasks.Task:FinishContinuations()
2023/04/21 16:17:29.567 6307 6324 Error Unity System.Threading.Tasks.Task`1:TrySetResult(TResult)
2023/04/21 16:17:29.567 6307 6324 Error Unity System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1:SetResult(TResult)
2023/04/21 16:17:29.567 6307 6324 Error Unity GLTFast.<LoadGltfBinary>d__53:MoveNext()
2023/04/21 16:17:29.567 6307 6324 Error Unity System.Threading.ContextCallback:Invoke(Object)
2023/04/21 16:17:29.567 6307 6324 Error Unity System.Threading.ExecutionContext:RunInternal(ExecutionContext, ContextCallback, Object, Boolean)
2023/04/21 16:17:29.567 6307 6324 Error Unity System.Runtime.CompilerServices.MoveNextRunner:Run()
2023/04/21 16:17:29.567 6307 6324 Error Unity System.Action:Invoke()
2023/04/21 16:17:29.567 6307 6324 Error Unity System.Threading.ContextCallback:Invoke(Object)
2023/04/21 16:17:29.567 6307 6324 Error Unity System.Threading.Tasks.AwaitTaskContinuation:Ru

Can someone kindly teach me how to solve this issue? 

Also there's still no controller visual for Vive Focus 3 ?
A fallback model is being loaded when I run the scene on Vive Focus 3. Currently, I replace the fallback model with the one I found in Wave SDK but I wonder if this is the right way to do it?

**Comment by keveleigh** [comment]
For the pink textures, do you have both the glTFast package and its KTX extension installed? https://github.com/atteneder/KtxUnity

For the animations, that's currently by-design. Meta implemented their controller model animations differently than MSFT, and MRTK3 doesn't currently support that implementation.

**Reply by s871922003** [reply]
Hi keveleigh,
Thanks for replying~ I actually have both gITFast and KTX extension installed in project, the strange thing is pink textures does not happens in HandInteractionSample scene so that scene must have something to handle the controller texture right?

And for the animations, sorry my native language is not english but do you mean if I want animation for the controllers I needed to let Oculus SDK to handle this part rather than MRTK3?
